[FEATURE] Bloquer les boutons quand le temps est écoulé (PIX-1454).

MelanieMEB opened the pull request:
## :unicorn: Problème
Les utilisateurs ne voient pas toujours que l'épreuve était en timeout

## :robot: Solution
Quand le temps imparti est écoulé : 
- On bloque la modification de la réponse
- On affiche seulement le bouton poursuivre

## :rainbow: Remarques

## :100: Pour tester
https://app-pr2034.review.pix.fr/challenges/receQkwO1dvjQc2S3/preview

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2034.review.pix.fr
- Orga: https://orga-pr2034.review.pix.fr
- Certif: https://certif-pr2034.review.pix.fr
- Admin: https://admin-pr2034.review.pix.fr
- API: https://api-pr2034.review.pix.fr/api/

 Please check it out!

MelanieMEB committed:
Add isTimeoutChallenge tracked variable for all challenge

MelanieMEB committed:
Block button and answer when the time is out for qroc challenge

- Add a new Font Icon
- Bloc button when the time is out
- Bloc answer when the time is out
- Change message when the user go again on this challenge

MelanieMEB committed:
Add block timeout challenge in qrocm challenge

MelanieMEB committed:
Add block timeout challenge in qcu challenge

MelanieMEB committed:
Add block timeout challenge in qcm challenge

Commit:
Add challenge actions tests

sbedeau commented on mon-pix/tests/integration/components/challenge-actions-test.js:
```diff
@@ -84,4 +84,23 @@ describe('Integration | Component | challenge actions', function() {
+  describe('Challenge is timedout', function() {
+
+    it('should only display "continue" button', async function() {
+      // when
+      this.set('isValidateButtonEnabled', true);
+      this.set('isTimeoutChallenge', true);
+      this.set('isSkipButtonEnabled', true);
+      this.set('externalAction', () => {});
+      await render(hbs`{{challenge-actions
+        validateAnswer=(action externalAction)
+        isValidateButtonEnabled=isValidateButtonEnabled
+        isTimeoutChallenge=isTimeoutChallenge
+        isSkipButtonEnabled=isSkipButtonEnabled}}`);
+
+      expect(findAll('.challenge-actions__action').length).to.equal(1);
+      expect(find('.challenge-actions__action-continue')).to.exist;
```
```suggestion
	  // given
      this.set('isValidateButtonEnabled', true);
      this.set('isTimeoutChallenge', true);
      this.set('isSkipButtonEnabled', true);
      this.set('externalAction', () => {});
      
      // when
      await render(hbs`{{challenge-actions
        validateAnswer=(action externalAction)
        isValidateButtonEnabled=isValidateButtonEnabled
        isTimeoutChallenge=isTimeoutChallenge
        isSkipButtonEnabled=isSkipButtonEnabled}}`);

	  // then
      expect(findAll('.challenge-actions__action').length).to.equal(1);
      expect(find('.challenge-actions__action-continue')).to.exist;
```

sbedeau commented on mon-pix/translations/en.json:
```diff
@@ -372,6 +372,9 @@
                 "qrocm": "To validate, please complete all answer fields. Otherwise, click Skip."
+            },
+            "timed" : {
+                "cant-answered": ""
```
```suggestion
                "cannot-answer": ""
```

sbedeau commented on mon-pix/translations/fr.json:
```diff
@@ -372,6 +372,9 @@
                 "qrocm": "Pour valider, veuillez remplir tous les champs réponse. Sinon, passez."
+            },
+            "timed" : {
+                "cant-answered": "Vous n'avez pas pu répondre dans le temps accordé."
```
```suggestion
                "cannot-answer": "Vous n'avez pas pu répondre dans le temps accordé."
```
ou
```suggestion
                "timedout": "Vous n'avez pas pu répondre dans le temps accordé."
```

sbedeau commented on mon-pix/app/components/challenge-item-generic.js:
```diff
@@ -66,6 +67,9 @@ export default class ChallengeItemGeneric extends Component {
       const timer = later(this, function() {
```
**suggestion (non-blocking):** Peut-être stopper le tick quand `this.isTimeoutChallenge = true` pour éviter des surprises

sbedeau commented on mon-pix/app/styles/components/_challenge-actions.scss:
```diff
@@ -149,6 +149,15 @@
   margin-right: 20px;
 }
 
+.challenge-actions__time-out {
```
```suggestion
.challenge-actions__timed-out {
```

sbedeau reviewed (approved):
Globalement ok techniquement

Fonctionnellement, j'ai remarqué que : 
- on peut continuer d'écrire même après l'état timed-out et la réponse est enregistrée (je m'attendais à ne pas pouvoir écrire du tout)
- le tick continue de s'incrémenter dans la width de la classe (potentiellement consommateur de perfs locales sur les grands nombres mais on est dans le nitpicking)
- la barre gagnerait peut-être en lisibilité si elle avait un borderline (optionnel)

A voir si ces remarques correspondent à nos attentes
<img width="743" alt="Screenshot 2020-10-26 at 18 22 18" src="https://user-images.githubusercontent.com/46371288/97206889-9ddfa380-17b9-11eb-8f51-5bf510f82e1e.png">
<img width="1002" alt="Screenshot 2020-10-26 at 18 22 06" src="https://user-images.githubusercontent.com/46371288/97206897-a041fd80-17b9-11eb-93f0-71e0fb6b55f5.png">
<img width="440" alt="Screenshot 2020-10-26 at 18 22 00" src="https://user-images.githubusercontent.com/46371288/97206908-a20bc100-17b9-11eb-9cea-d288d76dca6f.png">

jonasgrilleres commented on mon-pix/app/components/challenge-item-generic.js:
```diff
@@ -10,6 +10,7 @@ export default class ChallengeItemGeneric extends Component {
   @tracked isValidateButtonEnabled = true;
   @tracked isSkipButtonEnabled = true;
   @tracked hasUserConfirmedWarning = false;
+  @tracked isTimeoutChallenge = false;
```
```suggestion
  @tracked hasChallengeTimedOut = false;
```

jonasgrilleres commented on mon-pix/tests/integration/components/challenge-actions-test.js:
```diff
@@ -84,4 +84,26 @@ describe('Integration | Component | challenge actions', function() {
 
   });
+
+  describe('Challenge is timedout', function() {
```
```suggestion
  describe('Challenge has timed out', function() {
```

MelanieMEB committed:
Stop timer when it finished

MelanieMEB committed:
Rename isTimeoutChallenge to hasChallengeTimedOut

laura-bergoens reviewed (approved):
Ok, j'ai aussi testé fonctionnellement

PhilippineLoison commented:
Si je saisi une réponse sans valider et que je laisse le temps s'écouler, j'ai ce message "Vous n'avez pas pu répondre dans le temps accordé." Est ce qu'on devrait pas mettre un message plus général de type : 
"Le temps accordé est écoulé."
